[BUU] Activate admin_style_v3 for most system specs

filipefurtad0 opened the pull request:
#### What? Why?

!! Not to be merged just yet, as it activates BUU by default !!

- Required for https://github.com/openfoodfoundation/openfoodnetwork/issues/12627

It does not yet fully address feature parity, I've disconnected the issue [#11845](https://github.com/openfoodfoundation/openfoodnetwork/issues/11845)


<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->

- Enables the BUU toggle on `base_spec_helper.rb`;
- Fixes case on text assertions
- Fixes failing tests due to overlapping elements; a helper was created to deal with this

In what #11845 concerns, this is the current state (this will be updated as progressed):
```

As an Administrator
  I want to be able to manage products in bulk
  listing products
    displays a list of products - **DONE**
    displays a message when number of products is zero - **DONE**
    displays a select box for suppliers, with the appropriate supplier selected - **PENDING (should be fixed in #11932)**
    displays an on hand count in a span for each product - **DONE**
    displays 'on demand' for any variant that is available on demand - **DONE**
    displays a select box for the unit of measure for the product's variants - **PENDING**
    displays a text field for the item name when unit is set to 'Items' - **PENDING**

```
#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- System tests should display the BUU design on all admin pages
- Green build

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
This can be done once we have completed all BUU1 features, including:
* #12473 
* #12331 
* #12398 
* #11987 
* #11055 
* #12482 
* #12362 



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

filipefurtad0 commented:
Ok, some `/admin` specs are failing, moving to in Dev.

dacook commented:
Thanks Filipe! It looks like probably all of these failures are due to a style change in capitalisation, which results in text no longer matching. Eg:

```
  expected to find text "DISTRIBUTOR ..." in "Distributor..."
```

I'm not sure of the most efficient way to solve this, but I've often thought that these expectations shouldn't be case-sensitive (they still convey the same meaning to the user), so I would argue that we change that. I'll raise in #core-devs.

dacook commented:
I [discussed with Maikel](https://openfoodnetwork.slack.com/archives/GK2T38QPJ/p1696977666895229), and agreed it's worth updating the text in the specs to be more readable and better match what's on the screen.

We also agreed that it's helpful to optionally add a matcher like [`have_content_i`](https://stackoverflow.com/a/44575326/421243) to better handle this in the future. 

@filipefurtad0 would you like to try this, or perhaps hand over for a `good-first-issue`?

filipefurtad0 commented:
Happy to have a go at this :+1: 
Thanks @dacook

dacook commented on spec/system/admin/subscriptions/crud_spec.rb:
```diff
@@ -128,7 +128,7 @@
                 find("a.cancel-order").click
               end
-              expect(page).to have_content 'CANCELLED'
+              expect(page).to have_content 'CANCELLED'.upcase
```
🧐

filipefurtad0 replied:
Oops!

dacook reviewed (changes requested):
Thanks Filipe

Since you started this, the `/admin/products` URL now points to the new screen, so the old `bulk_product_update_spec` now fails (it's loading the new one instead of the old one). I think you should be able to fix this by disabling the feature toggle on that test.

filipefurtad0 commented:
Hum, indeed there are some failing tests now. I've left this one forgotten in In Dev, but now of course the code has changed (after rebasing).

Either way, even after fixing, I guess we'll only be able to merge this PR once we're ready to release the feature, right? So, maybe we should hold this one for a while... And update everything once we're ready to release BUU.

What do you think @dacook ?

dacook commented:
🤔 Hmm. 

* We want to continue supporting the old (current) admin styles, so need to ensure it's being tested
* We want to also test with the new styles, in case any new problems come up

We could run the specs twice, for both old and new. 
But I don't think it's necessary. Apart from the bulk products screen, the only changes are styling (colours, padding, fonts, etc). 

So I agree, let's keep this PR in dev, and wait until we get closer to releasing admin_style_v3. We can occasionally refresh this branch when needed until then.

filipefurtad0 commented:
> So I agree, let's keep this PR in dev, and wait until we get closer to releasing admin_style_v3. We can occasionally refresh this branch when needed until then.

Ok, great - thanks. I've updated it with a few changes so it's green again.

filipefurtad0 commented:
Ok, I've fixed conflicts, rebased, updated a test - it's green again.

It seems to me the challenge now, is to migrate the relevant [tests from the legacy products page](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/admin/bulk_product_update_spec.rb) into the spec for [new product page](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/admin/products_v3/products_spec.rb); After that, we can delete the legacy spec file.

Does this sound like a good plan @dacook?

dacook commented:
Awesome, thanks Filipe! I've created an issue for this: 
- https://github.com/openfoodfoundation/openfoodnetwork/issues/11845

I think we shouldn't delete the legacy spec until we remove the feature toggle.

dacook commented on spec/support/tom_select_helper.rb:
```diff
@@ -0,0 +1,13 @@
+      find('.ts-control').send_keys(value)
+    end
+
+    all('.ts-dropdown .ts-dropdown-content .option', text: /#{Regexp.quote(value)}/i)[0].click
```
Here `all` is potentially selecting multiple options (eg a value of "Distributor" could select "Distributor1" and "Distributor2"), then clicking on the first one. 
I think what we want is to use `find`, and I'd suggest against using Regexp if it's not needed. We would be looking for an exact match.

dacook commented on spec/support/tom_select_helper.rb:
```diff
@@ -0,0 +1,13 @@
+    container = find(:id, from)
+
+    within(container) do
+      find('.ts-control').send_keys(value)
```
I'm wondering if this is necessary. Doesn't capybara automatically scroll and find the desired element without the need for searching?

If needed, I wonder does the simpler `fill_in '.ts-control' with: value`  work?

rioug replied:
It does currently, but once we move to the latest version of tom select using `fill_in '.ts-control' with: value` will break.

filipefurtad0 replied:
So, I guess we leave this one as is?

PS: I also could not make it work when using `fill_in`:

```
     Failure/Error: fill_in '.ts-control', with: value
     
     Capybara::ElementNotFound:
       Unable to find field ".ts-control" that is not disabled within #<Capybara::Node::Element tag="div" path="//HTML[1]/BODY[1]/DIV[1]/DIV[3]/DIV[1]/DIV[1]/DIV[1]/FORM[1]/DIV[1]/FIELDSET[1]/DIV[1]/DIV[1]/DIV[1]">
```

rioug commented on spec/support/tom_select_helper.rb:
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module TomSelectHelper
```
There are already some helper for tom select here :
https://github.com/openfoodfoundation/openfoodnetwork/blob/dd502f0883d8099edec197d3368eb5d9420669ea/spec/support/request/web_helper.rb#L93
But like the idea of moving them to their own helper.

filipefurtad0 replied:
Ahh, I was unaware of those existing helpers... I'll move them to this file. The build should actually stay green, as we're calling the new file from `base_spec_helper`...

dacook reviewed (commented):
Looks ok to me. I have some ideas on simplifying the code, but it's possible that this code was necessary to get it working.

I'd love to see official helpers for this stuff, but sadly they [don't exist yet](https://github.com/orchidjs/tom-select/discussions/71#discussioncomment-5844625). Maybe one day we'll be able to contribute them :D

filipefurtad0 commented:
Thank you so much @dacook and @rioug for your early review on this :pray: 
If these two commits are both green then I'll squash them, and carry on with adding tests to this branch, to address #11845.

filipefurtad0 commented:
Hum, sorry - I'm a bit confused:
it seems both `select` (angular?) and `tom-select` (stimulus?) selecting working for the same field (highlighted in the pic below)... Both these approaches seem to work:

`select_tom_select supplier.name, from: 'product_supplier_field'`

`select supplier.name, from: 'product_supplier_id'`

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/8a0d78f9-62a3-48df-8c7b-2ff8f40f84fe)

Is this expected?

rioug commented:
> Is this expected? (I'd expect `select` not to work anymore - on the fields marked in green)

`select` is the Capybara one ([doc](https://www.rubydoc.info/github/jnicklas/capybara/Capybara/Node/Actions%3Aselect)) :

> #select(value = nil, from: nil, **options) ⇒ [Capybara::Node::Element](https://www.rubydoc.info/github/jnicklas/capybara/Capybara/Node/Element)
If from option is present, [#select](https://www.rubydoc.info/github/jnicklas/capybara/Capybara/Node/Actions#select-instance_method) finds a select box, or text input with associated datalist, on the page and selects a particular option from it. Otherwise it finds an option inside current scope and selects it. If the select box is a multiple select, [#select](https://www.rubydoc.info/github/jnicklas/capybara/Capybara/Node/Actions#select-instance_method) can be called multiple times to select more than one option. The select box can be found via its name, id, [test_id](https://www.rubydoc.info/github/jnicklas/capybara/Capybara#configure-class_method) attribute, or label text. The option can be found by its text.

I am not sure if it's expected to work with "tom select", but if it does I would be inclined to just use it. The git history doesn't really gives us context as to why the tom select helper were added, presumably because `select` didn't work, it's possible that newer version of "tom select" allowed `select` to work.

filipefurtad0 commented:
I have a question about the tom-select dropdowns. After a recent rebase, we seem to be getting several failing specs. I don't know where these changes were introduced... Any ideas? @rioug we've chatted a bit about this in delivery circle, but I'm not sure this is what you meant.

I've tried fixing some of them, like on this [commit](https://github.com/openfoodfoundation/openfoodnetwork/pull/11645/commits/54a25861c4d6e6035e26b3925486d94c5eb54d6e), but it's seems challenging to use the same helper for different drop-downs. Is this expected?

Now I've moved to .spec/admin/order_spec.rb (failing [here](https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/7833846887/job/21375799106?pr=11645).) The naming convention seems to vary bit. Using some existing helpers trigger this error:

```

Capybara::Cuprite::MouseEventFailed: Firing a click at coordinates [481.6937503814697, 331.3999938964844] failed. Cuprite detected another element with CSS selector "html.webkit.chrome.chrome121.linux.js.orientation_landscape.maxw_1280 body.admin.stimulus-reflex-disconnected div.ng-scope div#wrapper div#content.with-sidebar div.container div.row div.twelve.columns div#select-customer fieldset.no-border-bottom div.ts-wrapper.primary.single.plugin-dropdown_input.has-options.focus.input-active.dropdown-active div.ts-dropdown.single.plugin-dropdown_input div.dropdown-input-wrap input.dropdown-input" at this position. It may be overlapping the element you are trying to interact with. If you don't care about overlapping elements, try using node.trigger("click").
from /home/ffurtado/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/cuprite-0.15/lib/capybara/cuprite/page.rb:109:in `rescue in before_click'
Caused by Ferrum::JavaScriptError: Error: click, html.webkit.chrome.chrome121.linux.js.orientation_landscape.maxw_1280 body.admin.stimulus-reflex-disconnected div.ng-scope div#wrapper div#content.with-sidebar div.container div.row div.twelve.columns div#select-customer fieldset.no-border-bottom div.ts-wrapper.primary.single.plugin-dropdown_input.has-options.focus.input-active.dropdown-active div.ts-dropdown.single.plugin-dropdown_input div.dropdown-input-wrap input.dropdown-input, 481.6937503814697, 331.3999938964844
    at Cuprite.mouseEventTest (<anonymous>:296:11)
    at HTMLDivElement.<anonymous> (<anonymous>:1:31)
from /home/ffurtado/.rbenv/versions/3.1.4/lib/ruby/gems/3.1.0/gems/ferrum-0.14/lib/ferrum/frame/runtime.rb:138:in `handle_error'
```

Indeed, there are overlapping elements (**orders creation page**):

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/9f8fcf4c-585b-4b76-be50-7897e3faf1d4)

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/7b2c4d39-b46d-4390-913c-d7c48d8772f3)


**Product creation page**:

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/674b407b-6810-4b66-885a-49807a4cc1ec)

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/75f95e07-bef3-4cb2-a03c-6be0569f7140)

I'm unsure how to proceed. Is there a convention we should use:
- to name tom-select dropdowns?
- to use a given helper?

filipefurtad0 commented:
Since we cannot merge this branch and use it during development, I wonder if it is a better idea just to update this branch, shortly before full release (as discussed [here](https://openfoodnetwork.slack.com/archives/C01T75H6G0Z/p1707339007851579), we're looking at about 3 months time). Thoughts?

rioug commented:
> I have a question about the tom-select dropdowns. After a recent rebase, we seem to be getting several failing specs. I don't know where these changes were introduced... Any ideas? @rioug we've chatted a bit about this in delivery circle, but I'm not sure this is what you meant.
> 

Looks like some of the changes from master got lost, see https://github.com/openfoodfoundation/openfoodnetwork/blob/c0adf947124af8a763c691299acb5bb261536112/spec/support/request/web_helper.rb#L100-L106 , specifically the `send_keys` on line 104. It's missing here : 
https://github.com/filipefurtad0/openfoodnetwork/blob/54a25861c4d6e6035e26b3925486d94c5eb54d6e/spec/support/tom_select_helper.rb#L15-L21

I had a look at the first failing test, this one rspec ./spec/system/admin/order_spec.rb:475 
There are several things going on here, :
* the helper is called with the wrong parameter, it should be : `tomselect_search_and_select customer.email, from: 'customer_search_override'` 
* the helper is missing my fix from master, as shown above.


> I'm unsure how to proceed. Is there a convention we should use:
> - to name tom-select dropdowns?

I believe we just use rails convention here but it does seem a bit inconsistent

`tomselect_multiselect`, `tomselect_search_and_select`, `tomselect_select` all expect the name of the select component, ie: 

![tomselect_name](https://github.com/openfoodfoundation/openfoodnetwork/assets/40413322/4b3201b3-16e3-4108-92ef-bfe0bb66c429)

`select_tom_select` on the other hand expect an html id, and will go look for tomselect in the component's children.

> - to use a given helper?

I don't know the history of these helpers, but it looks like `tomselect_search_and_select` and `select_tom_select` are doing the same thing. So I would suggest to keep only one of these two and keeping the naming consistent. I added `tomselect_select` because sometimes tomselect is setup as a normal select, ie you see the options dropdown is already populated when you click on tomselect. Even though the search functionality is present, it not necessary to search to find your your option. In this scenario `tomselect_search_and_select` was breaking due to some race condition. See:  https://github.com/openfoodfoundation/openfoodnetwork/commit/98545741e6069a2ec288552bac7de6e4e690a450
So I would say use: 
* `tomselect_search_and_select` when searching is actually required, ie: the order spec example above.
* `tomselect_select` when searching is not required
I am not sure about `tomselect_multiselect` though.

Hopefully that should unblock you !

filipefurtad0 commented:
Thanks @rioug !

> Looks like some of the changes from master got lost, see 

Indeed. This might have been to a bad rebase (on me :see_no_evil: )
At some point in the discussion above we've discussed moving helpers to their own file. I've now added that change on the respective helper which was indeed lost. 

> the helper is called with the wrong parameter, it should be : tomselect_search_and_select customer.email, from: 'customer_search_override'
    the helper is missing my fix from master, as shown above.

Having this done, brings the build back to green - which is great. I was a bit lost here.


>    tomselect_search_and_select when searching is actually required, ie: the order spec example above.
    tomselect_select when searching is not required
    I am not sure about tomselect_multiselect though.

Agree to remove `select_tom_select` if we don't need it, but the thing is, I can't seem to make any of the three above helpers work for the Unit Size dropdown, all search for the `name` attribute, and there is none defined within `select` element:


![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/6c4a814a-e7de-48eb-bed4-0ff1ba9a9622)


Hence, I've created this additional helper the `select_tom_select`, which - as you say looks for the `id:`:

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/4027c2f8-8700-443f-bda0-4df1b215fada)

I guess we can only get rid of `select_tom_select`, once we add a name attribute to that `select` element. Not sure I'm missing something - what do you think?

dacook commented:
Hi Filipe, I've updated the `tomselect_*select` methods in this PR to be more flexible:
https://github.com/openfoodfoundation/openfoodnetwork/pull/12183/commits/520482b8b69ca5e5de618235bce80df9bbabd998

Once it's merged, you'll have the option of using the label text (just like `fill_in`), which I think is generally better for system specs.
Eg: 
```ruby
        tomselect_select "Weight (g)", from: "Unit Size"
```

dacook commented:
Hi @filipefurtad0 , I went ahead and rebased this to prepare for merge. I condensed some of the past commits, and added some more fixes. 
I also added some "TODO" comments where we are still referring to the old Bulk Edit Products screen. I'd like to migrate those over to the new screen soon, but I think it can wait so that we can merge this long-running branch.

dacook reviewed (approved):
Although the latest changes are mine, I'm approving to resolve the "changes requested".

rioug commented on spec/system/admin/product_import_spec.rb:
```diff
@@ -96,6 +96,8 @@
 
       wait_until { page.find("a.button.view").present? }
 
+      puts "TODO: migrate to v3"
```
Shouldn't that be tracked in an issue ?

dacook replied:
Yes, I'll add checkboxes in the above issue 👍

dacook replied:
Done

rioug reviewed (approved):
Looks good !

mkllnk commented on spec/system/admin/reports_spec.rb:
```diff
@@ -130,9 +131,14 @@
+=======
+      expect(page).to have_selector "#report-table table"
+      expect(page).to have_content "First Name Last Name Billing Address Email"
+>>>>>>> 0a69712a7b (Fixes case on assertions, on failing specs)
```
This looks like a mistake.

mkllnk replied:
But it seems to have been fixed later.

filipefurtad0 replied:
Yeah... Sorry. Indeed, this might have happened a couple of times. This branch lived for quite a long time and was fixed for conflicts/rebased countess times. I'm not sure if it would not have been a better approach to open this branch, at a later stage. Maybe a topic for discussion if we get to do a retro on BUU.

mkllnk commented on spec/system/admin/bulk_product_update_spec.rb:
```diff
@@ -29,12 +29,6 @@
       expect(page).to have_field "product_name", with: p2.name
     end
 
-    it "displays a message when number of products is zero" do
-      visit spree.admin_products_path
-
-      expect(page).to have_text "No products yet. Why don't you add some?"
-    end
-
```
The main commit message doesn't reflect the commit content.

mkllnk commented on spec/system/admin/order_cycles/complex_editing_multiple_updation_spec.rb:
```diff
@@ -35,6 +35,9 @@
       click_link 'Orders'
     end
 
+    # Click dismiss on distributor warning
+    click_button 'Dismiss'
```
Or we could setup the enterprise properly to avoid the warning?

filipefurtad0 replied:
Done in #12652 :pray:

mkllnk commented on spec/system/admin/order_cycles/simple_spec.rb:
```diff
@@ -531,7 +531,8 @@
 
         # When I save, any exchanges that I can't manage remain
-        click_button 'Save'
+        # overlapping warning, we need to use 'node.trigger("click")'
+        page.find(:button, "Save").trigger("click")
```
Can we dismiss the warning first? Otherwise this is a UX issue. Users can't "find button and trigger save" with their mouse.

mkllnk commented on spec/system/admin/order_spec.rb:
```diff
@@ -771,14 +775,22 @@ def new_order_with_distribution(distributor, order_cycle)
+=======
+          expect(page).to_not have_content different_shipping_method_for_distributor1.name
+          dismiss_warning
+>>>>>>> 2ddec2cff4 (Deals with overlapping elements)
```
Another conflict that could have been squashed.

mkllnk commented on spec/system/admin/products_v3/products_spec.rb:
```diff
@@ -165,18 +165,10 @@ def expect_other_columns_visible
+        expect(page).to have_content "On demand"
+        expect(page).to_not have_content "20" # does not display the total stock
+        expect(page).to have_content "16" # displays the stock for variant_2
+        expect(page).to have_content "4"  # displays the stock for variant_3
```
I think that the previous selectors were to avoid false passes of the spec. Imagine this spec running with _Enterprise 4_. It would show the number 4 even without displaying that stock.

filipefurtad0 replied:
Do you think we should revert to the xpath code? Maybe an alternative would be to include the count: attribute, or make the numbers a bit more specific, like 111, 222, 333 - I know, in theory, we could still have Enterprise 111, but maybe more unlikely.

What do you think @mkllnk @dacook ?

mkllnk replied:
Yes, I would go back to the xpath. If this was a more common use case then I would create helpers for that. Aren't there some helpers for accessing table rows by a displayed name, for example? That would be better.

dacook replied:
Yes, in fact the inputs have labels associated with them, so something this should work:
```
within row_containing_name("variant name") do
  expect(page).to have_field "On demand", checked: true
  expect(page).to_not have_field "On Hand", with: 20
end
```
(hmm, we need to fix that inconsistent capitalisation..)

dacook replied:
Sorry, that was targeting the fields inside the popout.

If we want to check the display, we can target the button (which opens the popout):
```
within row_containing_name("variant name") do
  expect(page).to have_button "On Hand", with: "On demand"
  expect(page).to_not have_button "On Hand", with: 20
end
```

dacook replied:
Even better might be to say something like this. But as far as I know Capybara doesn't have a convenient way to check all text (including button labels). We could make a helper for that, but it's not worth the bother..

```
within row_containing_name("variant name") do
  within column("On Hand") do
    expect(page).to have_text "On demand"
    expect(page).to_not have_text "20"
  end
end
```

mkllnk reviewed (approved):
Huge chunk of work. Well done for taming this.

There are many little bits I would have like to see improved in the history but this needs to be merged as soon as possible to avoid more rebase errors.
